test/wallet-http-test.js

```diff
@@ -1708,6 +1708,261 @@ describe('Wallet HTTP', function() {
       });
     });
   });
+
+  it('should reject malformed/invalid finish requests', async function() {
+    await assert.rejects(wclient.createBatchFinish('primary', {
+      passphrase: '',
+      names: ['invalid_finish_data']
+    }), /map must be a object./);
+
+    await assert.rejects(wclient.createBatchFinish('primary', {
+      passphrase: '',
+      names: [{name: 'domain_name'}]
+    }), /name and data must be present in every element./);
+
+    await assert.rejects(wclient.createBatchFinish('primary', {
+      passphrase: '',
+      names: [{name: 'domain_name', data: 'invalid data'}]
+    }), /data must be a object./);
+
+    await assert.rejects(wclient.createBatchFinish('primary', {
+      passphrase: '',
+      names: [{name: 'domain_name', data: {}}]
+    }), /data must be in form: {records: \[\]}/);
+  });
+
+  it('should redeem lost bid and register won bids', async function() {
+    const name1 = await nclient.execute('grindname', [6]);
+    const name2 = await nclient.execute('grindname', [6]);
+    const data = { records: [] };
+
+    await wclient.createBatchOpen('primary', {
+      names: [name1, name2],
+      passphrase: '',
+      broadcast: true,
+      sign: true
+    });
+
+    await mineBlocks(treeInterval + 1, cbAddress);
+
+    // wallet1(primary) wins name1, wallet2(secondary) wins name2
+    const wallet1Name1WinningBid = createBid(name1, 1000001, 'wallet-1-bid-1');
+    const wallet1Name2LosingBid = createBid(name2, 1000000, 'wallet-1-bid-2');
+    const wallet2Name1LosingBid = createBid(name1, 1000000, 'wallet-2-bid-1');
+    const wallet2Name2WinningBid = createBid(name2, 1000001, 'wallet-2-bid-2');
+
+    await wclient.createBatchBid('primary', {
+      passphrase: '',
+      bids: [wallet1Name1WinningBid, wallet1Name2LosingBid]
+    });
+
+    await mineBlocks(1, cbAddress);
+
+    await wclient.createBatchBid('secondary', {
+      passphrase: '',
+      bids: [wallet2Name1LosingBid, wallet2Name2WinningBid]
+    });
+
+    await mineBlocks(treeInterval + 1, cbAddress);
+
+    await wclient.createBatchReveal('primary', {
+      passphrase: '',
+      names: [name1, name2],
+      sign: true,
+      broadcast: true
+    });
+
+    await mineBlocks(1, cbAddress);
+
+    await wclient.createBatchReveal('secondary', {
+      passphrase: '',
+      names: [name1, name2],
+      sign: true,
+      broadcast: true
+    });
+
+    await mineBlocks(treeInterval + 1, cbAddress);
+
+    const wallet2Finish = await wclient.createBatchFinish('secondary', {
+      passphrase: '',
+      names: [{name: name1, data}, {name: name2, data}]
+    });
+
+    assert.deepStrictEqual(wallet2Finish.errorMessages, []);
+    assert.equal(wallet2Finish.processedFinishes.length, 2); // one redeem one finish
```

Comment:
It doesn't seem like these tests check if the correct bids were redeemed/regsitered, what do you think? Blockchain consensus will prevent actual errors, but if we have a problem in transaction construction and mess up which was the redeem and which was the register, then our REVEAL outputs will be locked in the bad, pending transaction and we might get a "could not resolve preferred inputs" error.

Reply:
yes good catch.

Reply:
I have updated the tests to assert for REDEEM and REGISTER values, please let me know if this would be enough or should we add additional checks ?